sources/web/datalab/polymer/components/datalab-files/datalab-files.ts

```diff
@@ -375,6 +375,84 @@ class FilesElement extends Polymer.Element {
   }
 
   /**
+   * This is the entry point for file upload functionality. This is here to avoid using
+   * the very ugly <input> element for file uploads.
+   */
+  _altUpload() {
+    this.$.altFileUpload.click();
+  }
+
+  /**
+   * Gets called after the user selected one or more files from the upload modal.
+   * For each of the selected files, reads its contents, converts it to base64, then
+   * uses the ApiManager to save it on the backend.
+   */
+  _upload() {
+    const inputElement = <HTMLInputElement>this.$.altFileUpload;
+    const files = inputElement.files;
+    const currentPath = this.currentPath;
+    const uploadPromises: Promise<any>[] = [];
+
+    if (files) {
+      for (let i = 0; i < files.length; ++i) {
+        const file = files[i];
+
+        // First, load the file data into memory.
+        const loadPromise = new Promise((resolve, reject) => {
+
+          const reader = new FileReader();
+
+          reader.onload = function() {
+            let itemData = reader.result;
+            // Extract the base64 data string
+            itemData = itemData.substr(itemData.indexOf(',') + 1);
+
+            const model: JupyterFile = {
+              name: file.name,
+              path: currentPath,
+              type: 'file',
+              format: 'base64',
+              content: itemData,
+            };
+
+            resolve(model);
+          }
+
+          // TODO: handle file reading errors.
+          reader.onerror = () => {
+            reject('Error reading file.');
+          }
+
+          // TODO: this will freeze the UI on large files (>~20MB on my laptop) until
```

Comment:
If there is a size beyond which the UI might freeze, I think we should limit the upload size to files smaller than that. You should be able to check file.size and reject the request if the file is too big.